### PR TITLE
Support syntax highlighting for more languages

### PR DESCRIPTION
**Pull request by lucifer1004**
Solves #768

### CODE REVIEW

**Comment by lucifer1004** [comment]
The problem is that I should not use the latest rulesets since `syntect` does not support Sublime Text's latest version yet.

**Comment by lucifer1004** [comment]
Use default features of `syntect` can solve the problem.

**Comment by laurmaedje** [comment]
Where does the `syntaxes.bin` come from?

**Comment by lucifer1004** [comment]
I compiled the binary syntax file using modified `bat` source code via `syntect`'s `dump` function.

**Comment by lucifer1004** [comment]
https://github.com/trishume/syntect/pull/470 has been merged, but we still need to wait for a new release of `syntect`.

File: library/Cargo.toml [hunk 1]
```diff
@@ -30,7 +30,7 @@ rustybuzz = "0.5"
 serde_json = "1"
 serde_yaml = "0.8"
 smallvec = "1.10"
-syntect = { version = "5", default-features = false, features = ["default-syntaxes", "regex-fancy"] }
+syntect = { version = "5" }
```
Comment (laurmaedje) [inline library/Cargo.toml:33]: why not `default-features = false` anymore?

**Comment by lucifer1004** [comment]
> Since Typst isn't on crates.io, we can also switch to the git version of syntect for now. However, it is unclear to me under what license terms we would use `syntaxes.bin` here.

For the license terms, we can refer to https://github.com/Canop/broot/tree/main/resources/syntect.

**Comment by lucifer1004** [comment]
The syntax file is now generated a bit differently to be compatible with the current dependency settings.

The following syntaxes are disabled due to compatibility issues:

- syntaxes/02_Extra/Assembly (ARM).sublime-syntax
- syntaxes/02_Extra/Elixir/Regular Expressions (Elixir).sublime-syntax
- syntaxes/02_Extra/JavaScript (Babel).sublime-syntax
- syntaxes/02_Extra/LiveScript.sublime-syntax
- syntaxes/02_Extra/PowerShell.sublime-syntax
- syntaxes/02_Extra/SCSS_Sass/Syntaxes/Sass.sublime-syntax
- syntaxes/02_Extra/SLS/SLS.sublime-syntax
- syntaxes/02_Extra/VimHelp.sublime-syntax
- syntaxes/02_Extra/cmd-help/syntaxes/cmd-help.sublime-syntax

Here the `syntaxes` directory refers to https://github.com/sharkdp/bat/tree/master/assets/syntaxes

Code for syntax set generation:

```rs
fn main() {
    let mut builder = syntect::parsing::SyntaxSet::load_defaults_nonewlines().into_builder();
    builder.add_from_folder("syntaxes/02_Extra", false).unwrap();
    let syntax_set = builder.build();
    syntect::dumps::dump_to_file(&syntax_set, "syntaxes.bin").unwrap();
}
```

**Comment by lucifer1004** [comment]
@reknih @laurmaedje Please review the changes.

**Comment by lucifer1004** [comment]
When generating the syntax set, `lines_include_newline = false` is used. According to https://github.com/trishume/syntect/blob/b5212417475f678706bdc243d7adbf99b26923bd/src/parsing/syntax_set.rs#L471, this might not work correctly for some syntaxes. But I have not found any examples (where this setting fails) yet.

However, if we use `lines_include_newline = true`, I did find some cases that the syntax highlighting cannot work correctly, e.g., the following Julia code,

```jl
# comment
function f() # comment
    2 # comment
end
```

will be completely rendered as comments.

This can be solved by tweaking https://github.com/typst/typst/blob/4b17c3f5c65a3ec87719d3c39d270767140e3767/library/src/text/raw.rs#L186 and adding a newline manually, but that would lead to other new issues. So I gave up and used `lines_include_newline = false` for now.

**Comment by lucifer1004** [comment]
In the future, I think we should design something like `bat`'s asset cache for both syntaxes and themes so that users can add new syntaxes or themes themselves.

**Comment by AlexanderMisel** [comment]
@lucifer1004 It seems we have to annotate the type as rust compiler suggests:

```bash
   --> library/src/text/raw.rs:314:5
    |
314 |     Lazy::new(syntect::parsing::SyntaxSet::load_defaults_nonewlines);
    |     ^^^^^^^^^ cannot infer type of the type parameter `T` declared on the struct `Lazy`
    |
help: consider specifying the generic arguments
    |
314 |     Lazy::<T, _>::new(syntect::parsing::SyntaxSet::load_defaults_nonewlines);
    |         ++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `typst-library` due to previous error
```

I changed the line to make it work:

```rust
Lazy::<syntect::parsing::SyntaxSet>::new(syntect::parsing::SyntaxSet::load_defaults_nonewlines);
```

**Comment by laurmaedje** [comment]
Thanks!

**Comment by MatejKafka** [comment]
Do you remember what was the issue with the PowerShell syntax file? I'd like to get PowerShell highlighting working in Typst for my thesis.

**Comment by peng1999** [comment]
> Do you remember what was the issue with the PowerShell syntax file? I'd like to get PowerShell highlighting working in Typst for my thesis.

You can find a fixed powershell syntax [here](https://discord.com/channels/1054443721975922748/1057632212671025162/1171264021031297044).

**Pull request merged**